lib/fauna/page.rb

```diff
@@ -0,0 +1,381 @@
+        @page_params == other.instance_variable_get(:@page_params) &&
+        fauna_map == other.fauna_map && ruby_map == other.ruby_map
+    end
+
+    alias_method :eql?, :==
+
+    # :section: Configuration
+
+    ##
+    # If the current page is populated.
+    #
+    # Unpopulated pages will not auto-populate; to populate the page, begin paging in any direction.
+    attr_reader :populated
+
+    # Client to execute queries with.
+    attr_reader :client
+
+    # The set query to paginate over.
+    attr_reader :set
+
+    # The configured ts to use for pagination.
+    def ts
+      @page_params[:ts]
+    end
+
+    ##
+    # The configured cursor to use for pagination.
+    #
+    # Automatically updated during pagination. The appropriate cursor is used based on paging direction.
+    def cursor
+      @page_params.select { |key, _| CURSOR_KEYS.include? key }
+    end
+
+    # The configured size to use for pagination.
+    def size
+      @page_params[:size]
+    end
+
+    # If pagination will include events.
+    def events
+      @page_params[:events]
+    end
+
+    # If pagination will include sources.
+    def sources
+      @page_params[:sources]
+    end
+
+    # An optional block used to apply a fauna map to the generated pagination query.
+    attr_reader :fauna_map
+
+    # An optional block used to map the data returned from the fauna query.
+    attr_reader :ruby_map
+
+    # :section: Data
+
+    ##
+    # Data contained within the current page.
+    #
+    # Always +nil+ for the initial page. Call one of the pagination methods to begin paging.
+    attr_reader :data
+
+    ##
+    # Before cursor for the current page.
+    #
+    # Always +nil+ for the initial page. Call one of the pagination methods to begin paging.
+    attr_reader :before
+
+    ##
+    # After cursor for the current page.
+    #
+    # Always +nil+ for the initial page. Call one of the pagination methods to begin paging.
+    attr_reader :after
+
+    # :section: Builders
+
+    ##
+    # Returns a copy of the page with the given +ts+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_ts(ts)
+      with_dup do |page|
+        page.instance_variable_get(:@page_params)[:ts] = ts
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given cursor set.
+    #
+    # Cursors either contain a +before+ or +after+ parameter.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_cursor(cursor = {})
+      with_dup do |page|
+        page.send(:load_cursor, cursor)
+      end
+    end
+
+    ##
+    # Returns a copy of the page with the given +size+ set.
+    #
+    # See {paginate}[https://faunadb.com/documentation/queries#read_functions-paginate_set] for more details.
+    def with_size(size = nil)
```

Comment:
calling might be more concise if size/ts/events/sources were set via a single with_params method.